src/python/WMCore/MicroService/Unified/MSRuleCleaner.py

```diff
@@ -528,30 +575,75 @@ def getMSOutputTransferInfo(self, wflow):
+            #       resolve the request (it is considered an ERROR outside WMCore)
+            if parentDataset is None:
+                msg = "Failed to resolve parent dataset for: %s in wflow: %s" % (childDataset, wflow['RequestName'])
+                raise MSRuleCleanerResolveParentError(msg)
+            elif parentDataset:
+                wflow['ParentDataset'].append(parentDataset[childDataset])
+                wflow['ParentDataset'] = list(set(wflow['ParentDataset']))
+                msg = "Found parent %s for input dataset %s in workflow: %s "
+                self.logger.info(msg, parentDataset, wflow['InputDataset'], wflow['RequestName'])
+            else:
+                msg = "Could not find parent for input dataset: %s in workflows: %s"
+                self.logger.error(msg, wflow['InputDataset'], wflow['RequestName'])
+        return wflow
+
     def getRucioRules(self, wflow, gran, rucioAcct):
         """
         Queries Rucio and builds the relevant list of blocklevel rules for
         the given workflow
         :param  wflow:   A MSRuleCleaner workflow representation
         :param  gran:    Data granularity to search for Rucio rules. Possible values:
-                        'block' || 'container'
+                         'block' or 'container'
         :return:         The workflow object
         """
         currPline = wflow['PlineMarkers'][-1]
-        # Find all the output placement rules created by the agents
-        for dataCont in wflow['OutputDatasets']:
-            if gran == 'container':
-                for rule in self.rucio.listDataRules(dataCont, account=rucioAcct):
-                    wflow['RulesToClean'][currPline].append(rule['id'])
-            elif gran == 'block':
-                try:
-                    blocks = self.rucio.getBlocksInContainer(dataCont)
-                    for block in blocks:
-                        for rule in self.rucio.listDataRules(block, account=rucioAcct):
-                            wflow['RulesToClean'][currPline].append(rule['id'])
-                except WMRucioDIDNotFoundException:
-                    msg = "Container: %s not found in Rucio for workflow: %s."
+
+        # Create the container list to the rucio account map and set the checkGlobalLocks flag.
+        mapRuleType = {self.msConfig['rucioWmaAccount']: ["OutputDatasets"],
+                       self.msConfig['rucioMStrAccount']: ["InputDataset", "MCPileup",
+                                                           "DataPileup", "ParentDataset"]}
+        if rucioAcct == self.msConfig['rucioMStrAccount']:
+            checkGlobalLocks = True
+        else:
+            checkGlobalLocks = False
+
+        # Find all the data placement rules created by the components:
+        for dataType in mapRuleType[rucioAcct]:
+            for dataCont in wflow[dataType] if isinstance(wflow[dataType], list) else [wflow[dataType]]:
```

Comment:
Isn't it possible that `wflow[dataType]` is an empty string? Thus getting into this for loop and failing with a rucio query for container `""`?

By the way, I think this line would be more readable if we break it down in multiple lines.

Reply:
No, it should never be an empty string. If it turns out to be such it means either the key value from the top level workflow representation coming from reqmgr is an empty string (which should never happen) or `findParent` call returned an empty string (which is also not possible) or some of the keys that we call this function with has a default value as `''` in `MSRuleCleanerWflow`, which is also not the case. BUT.... while speaking of that, I just realized (and tested it), we may end up with a `None` value instead of an empty string. Which I think could be equally bad regarding the Rucio call. But again if this happens I would say this is due to a different bug - the conditions under which it may happen is only if someone configures a workflow to have `IncludeParents` flag set to `True` and a value of `None` for `InputDataset`. I need to admit though haven't check if `StdBase` allows this.

Reply:
>if someone configures a workflow to have IncludeParents flag set to True and a value of None for InputDataset

there is a protection against it at the Specs level.

Reply:
And BTW, I see no easy way to protect the code here from such workflow configuration. What should I see do here? (this is supposed to be a rhetorical question ;) ) But joke aside, such a workflow configuration should be treated as a separate use case for sure, much further in the logic not here.

I hope I am not missing any other possibility. If you think of one ,now is the moment for us to take precautions.